src/main/lombok/eu/coinform/gateway/db/UserDbManager.java

```diff
@@ -66,6 +92,92 @@ public User logIn(String email, String password) throws AuthenticationException
         if (!passwordEncoder.matches(password, passwordAuth.get().getPassword())) {
             throw new BadCredentialsException("Incorrect Password");
         }
+
+        if(!userRepository.findById(passwordAuth.get().getId()).get().isEnabled()){
+            throw new UserNotVerifiedException("User not verified");
+        }
+
+        verificationTokenRepository.findByUser(passwordAuth.get().getUser()).ifPresent(token -> {
+            verificationTokenRepository.delete(token);
+        });
+
         return passwordAuth.get().getUser();
     }
+
+    public void createAndSaveVerificationToken(User user, String token){
+        VerificationToken myToken = new VerificationToken(token, user);
+        VerificationToken t = verificationTokenRepository.save(myToken);
+    }
+
+    public Optional<VerificationToken> getVerificationToken(String token){
+        return verificationTokenRepository.findByToken(token);
+    }
+
+    public Optional<VerificationToken> getVerificationToken(User user){
+        return verificationTokenRepository.findByUser(user);
+    }
+
+    public boolean confirmUser(String token) throws LinkTimedOutException{
+        Optional<VerificationToken> myToken = verificationTokenRepository.findByToken(token);
+        if(myToken.isEmpty()){
+            throw new NoSuchTokenException(token);
+        }
+        User user = myToken.map(VerificationToken::getUser).get();
+        if (myToken.get().checkExpiryDatePassed(new Date())) {
+            throw new LinkTimedOutException("Verification link timed out");
+        }
+        user.setEnabled(true);
+        userRepository.save(user);
+        verificationTokenRepository.delete(myToken.get());
+        return true;
+    }
+
+
+    public void logOut(Long userId){
+        Optional<User> user = userRepository.findById(userId);
+        user.ifPresent(u -> {
+            u.setCounter(u.getCounter()+1);
+            sessionTokenRepository.deleteById(userId);
+            userRepository.save(u);
+        }); // to invalidate the JWT token and remove longlived session
+    }
+
+    public String passwordReset(User user){
+        log.debug("Resetting user: {}", user.getPasswordAuth().getEmail());
+        user.setCounter(user.getCounter()+1); // to invalidate the JWT token
+        Optional<VerificationToken> oToken = verificationTokenRepository.findByUser(user);
+        String uuid = UUID.randomUUID().toString();
+
+        oToken.ifPresent(verificationToken -> verificationTokenRepository.delete(verificationToken));
```

Comment:
Ok. Changed so that while the token is valid it will not be changed mitigating this. If it is invalid we change it and send a new mail with the new link.